stdlib/public/core/FloatingPointRandom.swift

```diff
@@ -210,6 +174,448 @@ extension BinaryFloatingPoint where Self.RawSignificand: FixedWidthInteger {
+    let bExp = b.exponentBitPattern
+    
+    if a.sign == b.sign {
+      let sign = a.sign
+      let eSpan = (sign == .plus) ? (bExp - aExp) : (aExp - bExp)
+      if eSpan > 1 { return nil }
+      
+      let aSig = a.significandBitPattern
+      let bSig = b.significandBitPattern
+      let (low, high) = (sign == .plus) ? (aSig, bSig) : (bSig, aSig)
+      let x: Self
+      
+      if eSpan == 0 {
+        // Single raw binade
+        let s = RawSignificand.random(in: low..<high, using: &generator)
+        x = Self(sign: sign, exponentBitPattern: aExp, significandBitPattern: s)
+        
+      } else {
+        // Adjacent raw binades
+        let eBase = (sign == .plus) ? aExp : bExp
+        let isHigh: Bool
+        let s: RawSignificand
+        
+        if (eBase == 0) && (high <= low) {
+          // One subnormal
+          let span = high &+ (_significandMask &- low)
+          let r = RawSignificand.random(in: 0...span, using: &generator)
+          isHigh = r < high
+          s = isHigh ? r : low &+ (r &- high)
+          
+        } else if high <= (low &>> 1) {
+          // Both normal
+          let h2 = high &<< 1
+          let span = h2 &+ (_significandMask &- low)
+          let r = RawSignificand.random(in: 0...span, using: &generator)
+          isHigh = r < h2
+          s = isHigh ? (r &>> 1) : low &+ (r &- h2)
+          
+        } else {
+          // Large range
+          return nil
+        }
+        
+        let e = isHigh ? eBase + 1 : eBase
+        x = Self(sign: sign, exponentBitPattern: e, significandBitPattern: s)
+      }
+      
+      return (sign == .plus) ? x : x.nextDown
+      
+    } else if (aExp == 0) && (bExp == 0) {
+      // Subnormal opposite signs
+      let bSig = b.significandBitPattern
+      let span = a.significandBitPattern &+ bSig
+      if span < bSig { return nil }
+      
+      let r = RawSignificand.random(in: 0 ..< span, using: &generator)
+      let sign: FloatingPointSign = (r < bSig) ? .plus : .minus
+      let s = (r < bSig) ? r : r &- bSig &+ 1
+      return Self(sign: sign, exponentBitPattern: 0, significandBitPattern: s)
+      
+    } else {
+      // Large range
+      return nil
+    }
+  }
+  
+  // MARK: Helpers
+  
+  // If section numbers used 64 bits, then for ranges like `-1.0...1.0`, the
+  // `Int64.random(in:using:)` call in the general case would need to call
+  // `next()` twice on average. Each bit smaller than that halves the
+  // probability of a second `next()` call.
+  //
+  // The tradeoff is wider sections, which means an increased probability of
+  // landing in a section which spans more than one representable value and
+  // thus requires a second random integer.
+  //
+  // We optimize for `Double` by using 60 bits. This gives worst-case ranges
+  // like `-1.0...64.0` a 3% chance of needing a second random integer.
+  @_transparent
+  @_alwaysEmitIntoClient
+  internal static var _sectionBitCount: Int { UInt64.bitWidth - 4 }
```

Comment:
This will be unnecessary after #39143 “An optimal algorithm for bounded random integers” lands.

Reply:
May still be beneficial, let's measure; I have some further ideas for improving floating-point generation as well, though =)

Reply:
I look forward to hearing your ideas!